xds/internal/resolver/xds_resolver.go

```diff
@@ -236,94 +290,250 @@ func (r *xdsResolver) sendNewServiceConfig(cs *configSelector) bool {
+	// errors may occur.  Note: cs.clusters are pointers to entries in
+	// activeClusters.
+	for _, ci := range cs.clusters {
+		atomic.AddInt32(&ci.refCount, 1)
+	}
+
+	return cs, nil
+}
+
+// pruneActiveClusters deletes entries in r.activeClusters with zero
+// references.
+func (r *xdsResolver) pruneActiveClusters() {
+	for cluster, ci := range r.activeClusters {
+		if atomic.LoadInt32(&ci.refCount) == 0 {
+			delete(r.activeClusters, cluster)
 		}
 	}
 }
 
-// handleServiceUpdate is the callback which handles service updates. It writes
-// the received update to the update channel, which is picked by the run
-// goroutine.
-func (r *xdsResolver) handleServiceUpdate(su serviceUpdate, err error) {
-	if r.closed.HasFired() {
-		// Do not pass updates to the ClientConn once the resolver is closed.
+func (r *xdsResolver) addOrGetActiveClusterInfo(name string) *clusterInfo {
+	ci := r.activeClusters[name]
+	if ci != nil {
+		return ci
+	}
+
+	ci = &clusterInfo{refCount: 0}
+	r.activeClusters[name] = ci
+	return ci
+}
+
+type clusterInfo struct {
+	// number of references to this cluster; accessed atomically
+	refCount int32
+	// cfg is the child configuration for this cluster, containing either the
+	// csp config or the cds cluster config.
+	cfg xdsChildConfig
+}
+
+// Determines if the xdsResolver has received all required configuration from
+// the management server, and a matching virtual host was found in the route
+// configuration.
+func (r *xdsResolver) resolutionComplete() bool {
+	return r.listenerUpdateRecvd && r.routeConfigUpdateRecvd && r.currentVirtualHost != nil
+}
+
+// onResolutionComplete performs the following actions when resolution is
+// complete, i.e Listener and RouteConfiguration resources have been received
+// from the management server and a matching virtual host is found in the
+// latter.
+//   - creates a new config selector (this involves incrementing references to
+//     clusters owned by this config selector).
+//   - stops the old config selector (this involves decrementing references to
+//     clusters owned by this config selector).
+//   - prunes active clusters and pushes a new service config to the channel.
+//   - updates the current config selector used by the resolver.
+//
+// Only executed in the context of a serializer callback.
+func (r *xdsResolver) onResolutionComplete() {
+	if !r.resolutionComplete() {
 		return
 	}
-	// Remove any existing entry in updateCh and replace with the new one.
-	select {
-	case <-r.updateCh:
-	default:
+
+	cs, err := r.newConfigSelector()
+	if err != nil {
+		r.logger.Warningf("Failed to build a config selector for resource %q: %v", r.ldsResourceName, err)
+		r.cc.ReportError(err)
+		return
+	}
+
+	if !r.sendNewServiceConfig(cs) {
+		// JSON error creating the service config (unexpected); erase
+		// this config selector and ignore this update, continuing with
+		// the previous config selector.
+		cs.stop()
+		return
```

Comment:
I don't think either of these error conditions should hit, but is this what we want? Should these have two separate flows?

Reply:
> I don't think either of these error conditions should hit

Which error conditions are you talking about?

If you talking about the following two:
- `newConfigSelector`: this can fail when extracting the matcher from the route. See `xdsresource.RouteToMatcher`
- `sendNewServiceConfig`: this probably should never fail because we generate the service config in the xDS resolver. But defensive programming, yo!

Reply:
> Should these have two separate flows?

What do you mean by this?

Reply:
I agree with not calling cs.stop() since it's not there. I just think it should continue to use the old working configuration (this is an edge case that will never hit as it's written), as a JSON errors does. I'm not arguing changing the error plumbed back from route to matcher, just thinking it should have the same behavior as the JSON marshaling error, both errors that can never happen.

Reply:
I agree with not calling `cc.ReportError` when we cannot build a config selector. But I will be happy to make that change when we can guarantee that building the matchers from the provided route cannot fail, and the only way I can think of guaranteeing that is to build the matchers at resource unmarshal time and have it ready in the internal representation of the route. 

Now, we have two places where there are checks to see that the path matchers and header matchers are what we support, and although they seem to be doing the same thing today, there is no guarantee that these two implementations will not diverge tomorrow. So, it is better to do the conversion at unmarshal time, and that way, we can be dead sure that building a config selector will not fail.

Reply:
I feel like this is equivalent to an nack, which we will ignore and continue to use old configuration. "when we can guarantee that building the matchers from the provided route cannot fail" - I feel like what I'm arguing is to remove error, it's to treat this as an nack error and continue to use old configuration. Why should a user stop having his client work if we somehow don't nack but fail here, vs. an nack which will leave his configuration as such.